### DISCUSSION TITLE
Help with identifying and deleting leftover files after uninstalling HomeBrew

### DISCUSSION THREAD

**Discussion by ajdonim** [Homebrew/discussions#307] [Everyday usage] [answered]
I'm using a Macbook Pro 2018 on Mojave. I installed HomeBrew to run f3 to check some micro SD cards. I ran the uninstall program provided by HomeBrew, but it said it didn't uninstall everything. I received 2 notifications that files could not be deleted due to permission being denied. Which I'm not sure why since I'm on the admin account. At the end it also said some possible HomeBrew files were not deleted and that I may wish to remove them myself.

How do I delete the files that had permission denied? Also, how do I determine which of the possible files are indeed part of HomeBrew? Obviously /usr/local/.com.apple.installer.keep is not going to be part of HomeBrew, while /usr/local/Homebrew/ is. Though, how do I figure out the others? Also, how do I find the files? I copied and pasted the locations into Finder and nothing came up. Please explain in simple terms if possible. My background is in molecular biology, not computers.

<img width="327" alt="Screen Shot 2020-12-17 at 6 29 29 PM" src="https://user-images.githubusercontent.com/25938391/102569714-cdfd4180-409a-11eb-9e9d-e276b4ef98c8.png">

<img width="406" alt="Screen Shot 2020-12-17 at 6 29 54 PM" src="https://user-images.githubusercontent.com/25938391/102569721-d35a8c00-409a-11eb-8037-04e798c4a708.png">

**Comment by carlocab** [comment, marked as answer]
You can definitely remove `/usr/local/Homebrew` (as you've already noted) and `/usr/local/Caskroom` (if it's still there).

I believe you should also be able to delete `/usr/local/Frameworks`, but I'd want a second opinion on that.

I don't think `/usr/local/remotedesktop` belongs to Homebrew.

As for the rest, I don't think there is a simple systematic way of distinguishing Homebrew files from non-Homebrew files, unfortunately. The only way to do this for sure is to recall what you've installed into `/usr/local` on your own outside of Homebrew, and then remove the ones that don't fall into this category. (Not that helpful, I know.)

>How do I delete the files that had permission denied?

First, try `rm -f <file>`. If that doesn't work, then try `sudo rm -f <file>`. It will ask for your password. However, I would think twice about deleting something I need `sudo` to delete.

>how do I find the files?

You can do, for example

    cd /usr/local/Frameworks
    ls -la

to see the contents of `/usr/local/Frameworks`.

**Reply by ajdonim** [reply]
I'm very inexperienced with command prompts. My only experience was a short seminar at a workplace a few years ago and I've had no reason until now to use it, so forgot majority of it. From what I remember CD was to change the directory and I believe ls was to list out things. What is -la? 

How would I find /usr/local/Homebrew/ and /usr/local/Caskroom to delete them since they're not showing up using Finder? Do I have to use the command prompt to do so?

Also, do I want to delete /usr/local/Cellar? That was the other file where permission was denied.

**Reply by carlocab** [reply]
`ls` lists the contents of a directory. `-la` just tells it to list everything, including hidden files in a specific format. You can remove directories using, for example, `rm -rf /usr/local/Homebrew`. (Please be **very** careful with `rm -rf`.) You do want to delete `/usr/local/Cellar`.

If you prefer to do this in Finder:
1. Open a Finder window
2. Press Shift + Command + G
3. In the box that pops up, type `/usr/local`, then click "Go".

**Reply by ajdonim** [reply]
I used Finder and found Homebrew and Frameworks, but not Cellar or Caskroom. I deleted Homebrew.

I checked the contents of /usr/local/Frameworks using the command prompt as you directed and only 3 files appeared. They are named "." ".." and ".keepme". The file ".." shows as under root and wheel. Using Finder no files appear in the folder. Using the command prompt to see what's in Cellar and Caskroom showed the exact same as Frameworks.

How would you recommend I proceed from here?

**Reply by carlocab** [reply]
That means `/usr/local/Frameworks` contains only the `.keepme` file (you can always ignore the `.` and `..` entries). `.keepme` is a hidden file (hence why you need `-a`, and why it won't show up on Finder by default) that Homebrew puts in directories to avoid them being deleted in between OS upgrades.

You can delete `/usr/local/Frameworks`. Similarly, you can delete any empty folders (`ls -la` shows only `.` and `..`) or any folders that contain only `.keepme`.

>but not Cellar or Caskroom

This almost certainly means they've been deleted.

**Reply by ajdonim** [reply]
Ah ok, that makes sense. 

If Cellar and Caskroom not appearing under local using Finder means they've most likely been deleted, why are they showing up when I look into their directories using the command prompt?

**Reply by carlocab** [reply]
That's not possible. In the command prompt, please do `ls -la /usr/local` and paste the complete output here.

**Reply by ajdonim** [reply]
Looks like you're right. Neither showed up. Not sure then why they showed up when I looked into their directories and why Homebrew uninstaller said they were not deleted.

**Reply by carlocab** [reply]
It did not. The files/directories it said were not removed are here:

![](https://user-images.githubusercontent.com/25938391/102569721-d35a8c00-409a-11eb-8037-04e798c4a708.png)

**Reply by ajdonim** [reply]
Ok turns out looking into the directories actually wasn't working. I'm really tired and it seems I managed to skip over that it said there was no such file or directory and the files it was showing me were from Frameworks again since it couldn't change directories to one that didn't exist. Sorry about the confusion. Looks like I just need to remove Frameworks and I'm all finished. Thank you so much! I really appreciate you helping me so much.

**Reply by ajdonim** [reply]
> It did not. The files/directories it said were not removed are here:
> 
> ![](https://user-images.githubusercontent.com/25938391/102569721-d35a8c00-409a-11eb-8037-04e798c4a708.png)

Where I was confused is that it said it failed to delete them due to lack of permission. I thought that meant they were still there.

**Reply by ajdonim** [reply]
One more question. I started going through the other folders in local using Finder. If any of the folders have Homebrew files in them does that mean I can delete the folder? Or should I just delete the files? For a couple examples, the folder share (in local) contains multiple folders (doc, fish, man, zsh), and doc contains a file called homebrew. In local there's a folder called var which contains a folder called homebrew.
